lib/fibonacci.rb

```diff
@@ -4,6 +4,35 @@
 # e.g. 1st fibonacci number is 1
 # ....
 # e.g. 6th fibonacci number is 8
+
+# SOLUTION 1
+# My first solution has a time complexity of O(n) since the loop has to execute n-1
+# times. The time complexity is O(1) since it requires 3 additional variables
+# no matter the size of n.
 def fibonacci(n)
-  raise NotImplementedError
+  raise ArgumentError if n == nil || n < 0
+  return 0 if n == 0
+
+  num1 = 0
+  num2 = 1
+  num3 = 1
+  (n - 1).times do
+    num3 = (num1 + num2)
+    num1 = num2
+    num2 = num3
+  end
+  return num3
 end
+
+# SOLUTION 2
+# My second solution is a recursive one where the time complexity is O(n) and
```

Comment:
Nice work thinking about this recursively. It turns out that the space complexity of the recursive approach to fibonacci, like in your algorithm results in a time complexity of O(2^n). You already know that exponential is one of the worst time complexity. Also, the space complexity of this algorithm will be O(n) due to the call stack space it uses. We'll be taking a closer look to understand the time and space complexity of the recursive approach later in class.